src/cmd-tag

```diff
@@ -0,0 +1,177 @@
+
+    builds_json = "builds/builds.json"
+
+    build_data['timestamp'] = rfc3339_time()
+    write_json(builds_json, build_data)
+
+
+def available_builds(build_data):
+    """Returns the availble build IDs from build metadata"""
+
+    return build_data["builds"]
+
+
+def available_tags(build_data):
+    """Returns the available tag names from build metadata"""
+
+    return [t.get("name") for t in build_data.get("tags", {})]
+
+
+def cmd_list(args):
+    """List available tags in build metadata"""
+
+    build_data = init_build_data()
+    avail_tags = available_tags(build_data)
+    if not avail_tags:
+        print("No tags found")
+        return
+
+    for tag in build_data["tags"]:
+        print(json.dumps({"name": tag.get("name"),
+                          "created": tag.get("created"),
+                          "target": tag.get("target")}))
+
+
+def cmd_update(args):
+    """Create or update a tag to new build ID"""
+
+    build_data = init_build_data()
+    avail_tags = available_tags(build_data)
+    avail_builds = available_builds(build_data)
+
+    if "tags" not in build_data:
+        build_data["tags"] = []
+
+    # tag not available, so let's create it
+    if args.tag not in avail_tags:
+        # build not available, check for force
+        if args.build not in avail_builds:
+            # force not used, gotta bail
+            if not args.force:
+                fatal("Cannot create a tag with a build that does not exist")
+            if args.force:
+                print(f"INFO: Creating a new tag with a build ({args.build}) "
+                      "that does not exist")
+        # if we get here, either the build is available (and force can be
+        # ignored) or the build is not available and force has been used
+        created = rfc3339_time()
+        build_data["tags"].append({"name": args.tag,
+                                   "created": created,
+                                   "target": args.build})
+        finalize_json(build_data)
+        return
+
+    # tag should be available, so let's update it
+    # if the build isn't avaiable, check for force
+    if args.build not in avail_builds:
+        # force not used, bail out
+        if not args.force:
+            fatal("Cannot update a tag with a build that does not exist")
+
+        if args.force:
+            print(f"INFO: Updating the tag \"{args.tag}\" with a build "
+                  f"({args.build}) that does not exist")
+        # if we get here, either the build is available (and force can be
+        # ignored) or the build is not availble and force has been used
+        created = rfc3339_time()
+        build_data["tags"] = [{"name": args.tag,
+                               "created": created,
+                               "target": args.build}
+                              if t.get("name") == args.tag
+                              else t for t in build_data["tags"]]
+        finalize_json(build_data)
+        return
```

Comment:
Optional (since you've already respun this a bunch of times): we can dedupe this logic here, no? Something like:

```python
if args.build not in avail_builds:
  if not args.force:
    fatal(...)
  if args.force:
    print(...)

created = rfc3339_time()
if args.tag not in avail_tags:
  build_data["tags"].append(...)
else:
  build_data["tags"] = [{"name": ....}]
```

?

Reply:
Heh, I knew there was a better logic structure for this!  Thanks!  :+1: